src/garbage_collector.rs

```diff
@@ -220,42 +253,93 @@ impl GarbageCollectorOrchestrator {
 
     /// Run the garbage collector
     pub async fn run(&self) {
-        let manifest_ticker = Self::options_to_ticker(self.options.manifest_options.as_ref());
-        let wal_ticker = Self::options_to_ticker(self.options.wal_options.as_ref());
-        let compacted_ticker = Self::options_to_ticker(self.options.compacted_options.as_ref());
+        let manifest_schedule = self
```

Comment:
OK I updated the code, the most recent version of the code puts everything in the loop and uses the same tickers. The difference is that if it's a one-shot I just use `crossbeam_channel::at(Instant::now())` which just means that it immediately ticks once and never again.

At the same time, I return a mutable flag that indicates "how many more times it needs to run" (Indefinite/Once/Done), and after it runs once I set the flag to Done. If all three flags are done, I exit.

That way there's only one path to trigger the GC and that's in the `crossbeam_channel::select!`

----

The problem is that I'm trying to figure out how to prevent the `GarbageCollector` from being dropped, causing the channel to drop, before all this completes. For now I made `GarbageCollector` public but I don't think that's ideal.

Reply:
OK I fixed that last problem by shifting some code around and adding a shutdown watch channel.